server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java

```diff
@@ -388,30 +390,48 @@ public long dangerPermanentlyRemoveRecordsForGivenIDs(List<Long> ids) {
     private final DSLContext jooq;
     private final DSLContext readonlyJooq;
     private final ObjectMapper objectMapper;
-    private final SecretSeriesMapper secretSeriesMapper;
+    private final SecretSeriesMapper.SecretSeriesMapperFactory secretSeriesMapperFactory;
     private final RowHmacGenerator rowHmacGenerator;
-
-    @Inject public SecretSeriesDAOFactory(DSLContext jooq, @Readonly DSLContext readonlyJooq,
-        ObjectMapper objectMapper, SecretSeriesMapper secretSeriesMapper,
-        RowHmacGenerator rowHmacGenerator) {
+    private final GroupDAO.GroupDAOFactory groupDAOFactory;
+
+    @Inject public SecretSeriesDAOFactory(
+        DSLContext jooq,
+        @Readonly DSLContext readonlyJooq,
+        ObjectMapper objectMapper,
+        SecretSeriesMapper.SecretSeriesMapperFactory secretSeriesMapperFactory,
+        RowHmacGenerator rowHmacGenerator,
+        GroupDAO.GroupDAOFactory groupDAOFactory) {
       this.jooq = jooq;
       this.readonlyJooq = readonlyJooq;
       this.objectMapper = objectMapper;
-      this.secretSeriesMapper = secretSeriesMapper;
+      this.secretSeriesMapperFactory = secretSeriesMapperFactory;
       this.rowHmacGenerator = rowHmacGenerator;
+      this.groupDAOFactory = groupDAOFactory;
     }
 
     @Override public SecretSeriesDAO readwrite() {
-      return new SecretSeriesDAO(jooq, objectMapper, secretSeriesMapper, rowHmacGenerator);
+      return new SecretSeriesDAO(
+          jooq,
+          objectMapper,
+          secretSeriesMapperFactory.using(jooq),
+          rowHmacGenerator);
     }
```

Comment:
Integration testing revealed a subtle bug in some previous changes I'd made regarding automatically injecting a read-only instance of a GroupDAO into the SecretSeriesMapper. The symptom was that integration tests were timing out waiting for a database connection, and the integration tests run using a connection pool with a single connection. Specifically, they were timing out inside the SecretSeriesMapper while trying to look up the owner for secrets which had an owner. The root cause was that, in Jooq, a context (DSLContext) is roughly equivalent to a database connection, and every database query within the same context will share the same connection, more or less. By injecting a read-only GroupDAO up-front into the single, global SecretSeriesMapper we were essentially forcing Jooq to perform the owner lookup on a connection other than the one that was currently being used to look up the secrets. By introducing a factory for the SecretSeriesMapper we can scope it such that it shares the same context, and thus the same connection, of whatever query is currently executing.

Speaking of scoping, a lot (all?) of the DAO factory boilerplate would go away if we adopted Guice's child injectors. Basically, on every high-level database operation, create a child injector which binds DSLContext to the current context, and then inject DAOs etc. willy-nilly. I'm going to pursue this route in a follow-up PR but I didn't want to overcomplicate what is already a decent-sized change.